File: TwoPointer/287.find-the-duplicate-number.py
from typing import List


# @leet start
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = fast = 0
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                fast = 0
                while True:
                    if slow == fast:
                        return slow
                    slow = nums[slow]
                    fast = nums[fast]


# @leet end

File: TwoPointer/287.find-the-duplicate-number.go
package leetcode

// @leet start
func findDuplicate(nums []int) int {
	var slow, fast int
	for {
		slow, fast = nums[slow], nums[nums[fast]]
		if slow == fast {
			fast = 0
			for {
				slow, fast = nums[slow], nums[fast]
				if slow == fast {
					return slow
				}
			}
		}
	}
}

// @leet end
